src/main/scala/fi/aalto/cs/apluscourses/intellij/actions/REPLAction.scala

```diff
@@ -0,0 +1,64 @@
+package fi.aalto.cs.apluscourses.intellij.actions
+
+import com.intellij.execution.{RunManager, RunManagerEx, RunnerAndConfigurationSettings}
+import com.intellij.openapi.actionSystem.{AnActionEvent, CommonDataKeys}
+import com.intellij.openapi.module.{Module, ModuleUtilCore}
+import com.intellij.openapi.vfs.VirtualFile
+import org.jetbrains.annotations.{NotNull, Nullable}
+import org.jetbrains.plugins.scala.console.actions.RunConsoleAction
+import org.jetbrains.plugins.scala.console.configuration.ScalaConsoleRunConfiguration
+
+import scala.collection.JavaConverters._
+
+class REPLAction extends RunConsoleAction {
+
+  override def actionPerformed(e: AnActionEvent): Unit = {
+    customDoRunAction(e)
+  }
+
+  def checkFileOrFolderIsNull(@Nullable fileOrFolder: VirtualFile): Boolean = {
+    fileOrFolder == null
+  }
+
+  def adjustRunConfigurationSettings(@NotNull module: Module, @NotNull configuration: ScalaConsoleRunConfiguration): Unit = {
+    // adjust the configuration with: name, workDir and module
+    val moduleDirPath = ModuleUtilCore.getModuleDirPath(module)
+    configuration.setWorkingDirectory(moduleDirPath)
+    configuration.setModule(module)
+    configuration.setName("Scala REPL for module: " + module.getName)
+  }
+
+  def customDoRunAction(e: AnActionEvent): Unit = {
+    val dataContext = e.getDataContext
+    val project = CommonDataKeys.PROJECT.getData(dataContext)
+    // virtual file is working for both: files and folders
+    val targetFileOrFolder = CommonDataKeys.VIRTUAL_FILE.getData(dataContext)
+
+    if (project == null || checkFileOrFolderIsNull(targetFileOrFolder)) return
+
+    val runManagerEx = RunManagerEx.getInstanceEx(project)
+    val configurationType = getMyConfigurationType
+    val settings = runManagerEx.getConfigurationSettingsList(configurationType).asScala
+
+    // get target module
+    val module = ModuleUtilCore.findModuleForFile(targetFileOrFolder, project)
+
+    //choose the configuration to run based on the condition if this a new configuration of not
+    def chooseConfigurationSettings: RunnerAndConfigurationSettings = {
```

Comment:
created a nested function just to ask someone (@nikke234) what's the point of this language feature :wink:

Reply:
Hmm... An interesting question. I think nested functions are useful 1) when called multiple times, or, 2) for some recursive functions to keep the public interface cleaner (pseudo-code example):
```
int sum(list) {
  int sum_internal(s, list) {
    if (empty(list))
      return s;
    else
      return sum_internal(s + first(list), rest(list));
  }
  return sum_internal(0, list);
}
```

Reply:
Some helper functions or tail-recursive inner functions as in Olli's pseudo code are some good use cases. The public interface can be kept clean with `private`, but I think one advantage nested functions is the ability to access variables of the outer function:
```
def outer(x: Int): Int = {
  val y = x + 5
  def inner(a: Int): Int = {
    // Can read (and even modify) x and y here
  }
}
```

Reply:
Yes, I strongly agree with Nikolas's point about closures (i.e., helper functions "capturing" variables from the surrounding context). Once you get used to the idea of local, nested functions, this sort of thing is often handy and feels very natural.   

And yes, it's true that `private` is enough for keeping the public interface clean, but nested (i.e., _local_ functions) still add to that. This is in many respects analogous to how we use local variables even though we could use private variables instead. If we adopt the stance that functions are a primary means of abstraction just like variables are (i.e., functions are "first-class citizens" of the programming language, as people say), then it makes sense that functions, just like variables, can be restricted to a local scope.

Reply:
I see now, however, for OO languages, Classes (representing objects) often allow some way to see into them, and therefore unit test, whereas the idea of nested does not seem to provide this option